### DISCUSSION TITLE
Unsupported format 'qcow2'

### DISCUSSION THREAD

**Discussion by flombois** [bpg/terraform-provider-proxmox#1403] [Q&A]
Hi Everybody

I am trying to automate the deployment of a Ubuntu cloud VM as shown in the provider plugin documentation: [cloud-image](https://registry.terraform.io/providers/bpg/proxmox/0.60.0/docs/guides/cloud-image), I know Proxmox does not natively supports qcow2 format but the documentation mention that the provider can do the conversion however I am getting the following error when applying the configuration:

```shell
terraform apply --auto-approve -var-file=.tfvars
```

```
2024-06-20T16:50:57.478+0200 [ERROR] vertex "proxmox_virtual_environment_vm.vault_vm" error: error waiting for VM creation: task "UPID:backup0:000E4EC2:18A7E4674:667441D0:qmcreate:102:root#@pam!terraform:" failed to complete with exit code: unable to create VM 102 - unsupported format 'qcow2' at /usr/share/perl5/PVE/Storage/LVMPlugin.pm line 366.
╷
│ Error: error waiting for VM creation: task "UPID:backup0:000E4EC2:18A7E4674:667441D0:qmcreate:102:root@pam!terraform:" failed to complete with exit code: unable to create VM 102 - unsupported format 'qcow2' at /usr/share/perl5/PVE/Storage/LVMPlugin.pm line 366.
│ 
│   with proxmox_virtual_environment_vm.vault_vm,
│   on resources.tf line 45, in resource "proxmox_virtual_environment_vm" "vault_vm":
│   45: resource "proxmox_virtual_environment_vm" "vault_vm" {
│ 
╵
2024-06-20T16:50:57.500+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-06-20T16:50:57.500+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/bpg/proxmox/0.60.0/darwin_arm64/terraform-provider-proxmox_v0.60.0 pid=30083
2024-06-20T16:50:57.501+0200 [DEBUG] provider: plugin exited
```

I am running Proxmox 7.2

This is my terraform deployment configurations:

resources.tf
```terraform


# Ubuntu
locals {
  img_repository         = "https://cloud-images.ubuntu.com"
  img_os                 = "ubuntu"
  img_codename           = "noble"
  img_version            = "server"
  img_release            = "20240612"
  img_type               = "cloudimg"
  img_arch               = "amd64"
  img_format             = "img"
  img_checksum           = "784f6811c7681110cb066c8b4ad2a8fa654ad9c3ed63278286389f4180377600"
  img_checksum_algorithm = "sha256"
  img_filename           = "${local.img_codename}-${local.img_version}-${local.img_type}-${local.img_arch}.${local.img_format}"
  img_url                = "${local.img_repository}/${local.img_codename}/${local.img_release}/${local.img_filename}"
}

# Download the source image
resource "proxmox_virtual_environment_download_file" "img_source" {
  content_type       = "iso"
  datastore_id       = var.pve_node_img_storage_id
  file_name          = local.img_filename
  node_name          = var.pve_node
  url                = local.img_url
  checksum           = local.img_checksum
  checksum_algorithm = local.img_checksum_algorithm
}

# Create a proxmox vm and configure it through cloud init
resource "proxmox_virtual_environment_vm" "vault_vm" {
  depends_on  = [proxmox_virtual_environment_download_file.img_source]
  name        = "Vault"
  vm_id       = var.vm_id
  description = "Hashicorp vault instance managed by Terraform"
  tags        = ["terraform", "ubuntu", "hashicorp", "vault"]
  node_name   = var.pve_node

  startup {
    order = "3"
    #up_delay   = "60"
    #down_delay = "60"
  }

  agent {
    enabled = false
  }

  stop_on_destroy = true

  cpu {
    cores   = var.vm_cpu_cores
    sockets = var.vm_cpu_sockets
  }

  memory {
    dedicated = var.vm_memory
  }

  network_device {
    bridge = "vmbr1"
  }

  # System disk
  disk {
    interface    = "virtio0"
    datastore_id = var.vm_system_disk_datastore
    file_id      = proxmox_virtual_environment_download_file.img_source.id
    size         = var.vm_system_disk_size
  }

  # Data disk
  disk {
    interface    = "scsi0"
    datastore_id = var.vm_data_disk_datastore
    size         = var.vm_data_disk_size
  }

  initialization {
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }
  }

  operating_system {
    type = "l26"
  }
}
```

main.tf
```terraform
terraform {
  required_providers {
    proxmox = {
      source  = "bpg/proxmox"
      version = "0.60.0"
    }
  }
}

locals {
  api_token = "${var.pve_username}!${var.pve_token_id}=${var.pve_token_secret}"
}

provider "proxmox" {
  endpoint  = var.pve_endpoint
  api_token = local.api_token
  insecure  = var.pve_insecure

  ssh {
    agent    = true
    username = var.pve_ssh_username
    password = var.pve_ssh_password
  }
}
```

If I try to use file_format = 'raw' I am getting a different error:

```
2024-06-20T16:52:56.757+0200 [ERROR] provider.terraform-provider-proxmox_v0.60.0: Response contains error diagnostic: tf_proto_version=6.6 tf_rpc=ApplyResourceChange tf_resource_type=proxmox_virtual_environment_vm @caller=github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/internal/diag/diagnostics.go:58 @module=sdk.proto diagnostic_detail="" diagnostic_severity=ERROR
  diagnostic_summary=
  | resizing disk: error waiting for VM disk resize: All attempts fail:
  | #1: the server did not include a data object in the response
   tf_provider_addr=registry.terraform.io/bpg/proxmox tf_req_id=2802ca8b-e2b7-ae12-5324-546c9c390e09 timestamp="2024-06-20T16:52:56.757+0200"
2024-06-20T16:52:56.780+0200 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2024-06-20T16:52:56.780+0200 [ERROR] vertex "proxmox_virtual_environment_vm.vault_vm" error: resizing disk: error waiting for VM disk resize: All attempts fail:
#1: the server did not include a data object in the response
╷
│ Error: resizing disk: error waiting for VM disk resize: All attempts fail:
│ #1: the server did not include a data object in the response
│ 
│   with proxmox_virtual_environment_vm.vault_vm,
│   on resources.tf line 45, in resource "proxmox_virtual_environment_vm" "vault_vm":
│   45: resource "proxmox_virtual_environment_vm" "vault_vm" {
│ 
╵
2024-06-20T16:52:56.791+0200 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-06-20T16:52:56.794+0200 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/bpg/proxmox/0.60.0/darwin_arm64/terraform-provider-proxmox_v0.60.0 pid=30096
2024-06-20T16:52:56.794+0200 [DEBUG] provider: plugin exited
```

Any clue how I can enable the image format conversion ?

**Comment by bpg-dev** [comment]
Hi @flombois 👋🏼 

It _might_ be related to Proxmox 7.2 -- this version is not [officially supported](https://github.com/bpg/terraform-provider-proxmox?tab=readme-ov-file#compatibility-promise) by the provider, and I'm not testing against it. But I haven't look deeper just yet.

There is also a similar issue in #1404

**Comment by flombois** [comment]
Hi @bpg , thanks for your quick answer

I'll try to upgrade Proxmox as soon as possible to check whether the same issue happens again.

**Comment by flombois** [comment]
I just upgraded my Proxmox from 7.2 -> 7.4 -> 8.2 as mentioned in https://pve.proxmox.com/wiki/Upgrade_from_7_to_8#Continuously_use_the_pve7to8_checklist_script using 'in-place' upgrade. I tried to redeploy with the exact same config however I am running in the same exact issue.

However if I use file_format = 'raw' then it runs fine now and create the disks. Maybe my Proxmox instance is not properly configured to support 'qcow2' format. I'll continue investigating.

**Reply by bpg-dev** [reply]
The documentation for [`file_format`](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm#file_format) is not super clear, and we can probably improve it.

This parameter defines not the _source_ disk image file format, but the target format for the disk in the storage. And this depends on the storage type. Block storages, like LVM that you have, support only `raw` format, while directory-based storages allow you to select different formats, including `qcow2`.

Examples from the PVE UI:

Block storage:
<img width="606" alt="Screenshot 2024-06-21 at 9 41 11 PM" src="https://github.com/bpg/terraform-provider-proxmox/assets/627562/312b4575-52be-423e-8e61-336010fd78fa">

Directory storage:
<img width="598" alt="Screenshot 2024-06-21 at 9 42 56 PM" src="https://github.com/bpg/terraform-provider-proxmox/assets/627562/7225e417-1b89-4b03-879b-fa402189d935">

Hope this clarifies things a bit.

**Reply by flombois** [reply]
Thanks for those information, I'll check soon with different storage type and I think it should be the main issue indeed.

**Reply by ymartin59** [reply]
Hello,
Running Proxmox with Ceph RBD as storage we also faces the "resize" error message... until we enforce `file_format = raw` (instead of leaving defaults) as it is the only supported format for this storage type https://pve.proxmox.com/wiki/Storage:_RBD#ceph_rados_block_devices

**Reply by bpg-dev** [reply]
@ymartin59 was it on PVE 7.x as well?

**Comment by 1-cameron** [comment]
I'm having a similar issue with creating VMs with `qcow2` disks on Proxmox VE 7.4 (I've had no issue on PVE 8) and `0.55.0` of the provider:

```
Error: resizing disk: error waiting for VM disk resize: All attempts fail
```

However, with version `0.46.1` this is not a problem so that can be a workaround for anyone who's stuck on PVE 7.4. Did something change between these versions that would affect disks in this way?

I'll try with the latest version of the provider soon to see if the issue persists.

**Comment by bpg-dev** [comment]
Hi @1-cameron,

That error message was [added](https://github.com/bpg/terraform-provider-proxmox/pull/1336/files?diff=unified&w=0#diff-f6144f95e1ea6688c9a6244b06ccb8701e2fb8b4715ed3496bc0e65c0761e97eR438-R439) in version `0.58.1` as part of #1336. The change involved using the PVE [resize](https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/qemu/{vmid}/resize) API instead of custom pvesh commands over SSH. 
This API has [changed in 8.x](https://pve.proxmox.com/wiki/Roadmap#Proxmox_VE_8.0):
> When resizing a disk, spawn a worker task to avoid HTTP request timeout (issue 2315).

As [noted](https://github.com/bpg/terraform-provider-proxmox?tab=readme-ov-file#compatibility-promise) in the project's readme, the provider only supports the latest major version of PVE, so unfortunately, breakages with older versions will occur. I don't have the resources to test or maintain backward compatibility with PVE 7.x across the entire feature set.
